### PR TITLE
fix Miri visiting generators

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes fall-out caused by https://github.com/rust-lang/rust/pull/59897.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -262,8 +264,12 @@ impl<'rt, 'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>>
         variant_id: VariantIdx,
         new_op: OpTy<'tcx, M::PointerTag>
     ) -> EvalResult<'tcx> {
-        let name = old_op.layout.ty.ty_adt_def().unwrap().variants[variant_id].ident.name;
-        self.visit_elem(new_op, PathElem::Variant(name))
+        let name = match old_op.layout.ty.ty_adt_def() {
+            Some(def) => PathElem::Variant(def.variants[variant_id].ident.name),
+            // Generators also have variants but no def
```
Comment (eddyb) [inline src/librustc_mir/interpret/validity.rs]: Maybe pattern-match on the type explicitly?

File: src/librustc_mir/interpret/visitor.rs [hunk 2]
```diff
@@ -263,6 +252,13 @@ macro_rules! make_value_visitor {
                         // recurse with the inner type
                         return self.visit_field(v, 0, Value::from_mem_place(inner));
                     },
+                    ty::Generator(..) => {
+                        // FIXME: Generator layout is lying: it claims a whole bunch of fields exist
+                        // when really many of them can be uninitialized.
+                        // Just treat them as a union for now, until hopefully the layout
+                        // computation is fixed.
+                        return self.visit_union(v);
```
Comment (eddyb) [inline src/librustc_mir/interpret/visitor.rs]: cc @tmandry

File: src/librustc_mir/interpret/validity.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Done.

File: src/librustc_mir/interpret/validity.rs [hunk 3]
```diff
@@ -66,6 +66,7 @@ macro_rules! try_validation {
 pub enum PathElem {
     Field(Symbol),
     Variant(Symbol),
+    GeneratoreState(VariantIdx),
```
Comment (tmandry) [inline src/librustc_mir/interpret/validity.rs]: sp: `GeneratorState`

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: *oopsie* fixed in https://github.com/rust-lang/rust/pull/60569

**Pull request merged**